### PR TITLE
Refactor OrcWriterOptions to builder pattern

### PR DESCRIPTION
**Pull request by arunthirupathi**
OrcWriterOptions has 10 with* options to customize the OrcWriterOptions.
Each one of them changes one field, but copies the rest. So adding a new
option will require updating the existing 10 with* methods. There are few
additional options that are going to be introduced like SortDictionary Keys, 
StripeCache and FlatMapWriter.

Refactored the OrcWriterOptions to use builder pattern. With the new 
pattern any new options will require only OrcWriterOptions constructor,
builder, copy and build methods to be updated.

Test plan
Existing tests and added new tests for the new code path.

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-hive/src/main/java/com/facebook/presto/hive/OrcFileWriterConfig.java [hunk 1]
```diff
@@ -20,94 +20,96 @@
...
 {
-    private OrcWriterOptions options = new OrcWriterOptions();
+    private final OrcWriterOptions.Builder optionsBuilder = OrcWriterOptions.builder();
 
-    public OrcWriterOptions toOrcWriterOptions()
+    public OrcWriterOptions.Builder toOrcWriterOptionsBuilder()
     {
-        return options;
+        // toOrcWriterOptions could be called concurrently from multiple threads.
+        // Copy the builder for isolation and  thread safety.
```
Comment (sdruzkin) [inline presto-hive/src/main/java/com/facebook/presto/hive/OrcFileWriterConfig.java:28]: 1) Fix the double space before `thread safety.`
2) I'd rephrase  this in terms of immutability, and drop everything else about isolation and threads.

File: presto-hive/src/main/java/com/facebook/presto/hive/OrcFileWriterFactory.java [hunk 2]
```diff
@@ -124,7 +123,7 @@ public OrcFileWriterFactory(
             NodeVersion nodeVersion,
             DateTimeZone hiveStorageTimeZone,
             FileFormatDataSourceStats readStats,
-            OrcWriterOptions orcWriterOptions,
+            OrcFileWriterConfig orcFileWriterConfig,
```
Comment (sdruzkin) [inline presto-hive/src/main/java/com/facebook/presto/hive/OrcFileWriterFactory.java:126]: OrcFileWriterConfig is mutable. By replacing OrcWriterOptions with OrcFileWriterConfig you exposed this class to changes that might be made to OrcFileWriterConfig after passing it into the constructor and actually using in down below in the createFileWriter method.

I don't know about the lifecycle of OrcFileWriterConfig in Presto, but I'd not replace immutable object with a mutable version unless I have a good reasoning for it.

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcWriterOptions.java [hunk 3]
```diff
@@ -316,4 +151,184 @@ public String toString()
...
+        public Builder withStripeMaxRowCount(int stripeMaxRowCount)
+        {
+            checkArgument(stripeMaxRowCount >= 1, "stripeMaxRowCount must be at least 1");
+            this.stripeMaxRowCount = stripeMaxRowCount;
+            return this;
+        }
+
+        public Builder withRowGroupMaxRowCount(int rowGroupMaxRowCount)
+        {
+            checkArgument(rowGroupMaxRowCount >= 1, "stripeMaxRowCount must be at least 1");
```
Comment (sdruzkin) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcWriterOptions.java:228]: stripeMaxRowCount in the message -> rowGroupMaxRowCount

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcWriterOptions.java [hunk 4]
```diff
@@ -316,4 +151,184 @@ public String toString()
                 .add("integerDictionaryEncodingEnabled", integerDictionaryEncodingEnabled)
                 .toString();
     }
+
```
Comment (sdruzkin) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcWriterOptions.java:154]: See if it makes sense to also add `OrcWriterOptions.toBuilder`.

File: presto-orc/src/test/java/com/facebook/presto/orc/TestDictionaryColumnWriter.java [hunk 5]
```diff
@@ -608,7 +619,7 @@ private void verifyDwrfDirectEncoding(int stripeCount, List<StripeFooter> stripe
     private List<StripeFooter> testDictionary(Type type, OrcEncoding encoding, boolean enableIntDictionary, DirectConversionTester directConversionTester, List<?> values)
             throws IOException
     {
-        OrcWriterOptions orcWriterOptions = new OrcWriterOptions().withStripeMaxRowCount(STRIPE_MAX_ROWS).withIntegerDictionaryEncodingEnabled(enableIntDictionary);
+        OrcWriterOptions orcWriterOptions = OrcWriterOptions.builder().withStripeMaxRowCount(STRIPE_MAX_ROWS).withIntegerDictionaryEncodingEnabled(enableIntDictionary).build();
```
Comment (sdruzkin) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestDictionaryColumnWriter.java:622]: Would be nice to reformat this place too:
```java
OrcWriterOptions orcWriterOptions = OrcWriterOptions.builder()
  .withStripeMaxRowCount(STRIPE_MAX_ROWS)
  .withIntegerDictionaryEncodingEnabled(enableIntDictionary)
  .build();
```

File: presto-hive/src/main/java/com/facebook/presto/hive/OrcFileWriterConfig.java (hunk 1 above)
Comment (arunthirupathi) [reply presto-hive/src/main/java/com/facebook/presto/hive/OrcFileWriterConfig.java]: In my understanding using a copy is for isolation and less about immutability. Each user gets its own copy, so it has to do with isolation.

File: presto-hive/src/main/java/com/facebook/presto/hive/OrcFileWriterFactory.java (hunk 2 above)
Comment (arunthirupathi) [reply presto-hive/src/main/java/com/facebook/presto/hive/OrcFileWriterFactory.java:127]: This is a great point. Today OrcFileWriterConfig is used for config serialization and de-serialization and once deserialized it is immutable. So the current change is ok. 

But to future proof the code and to maintain the semantics, instead of capturing the OrcFileWriterConfig , I will capture the OrcWriterOptions.Builder which will provide the same level of isolation as the existing code.

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcWriterOptions.java (hunk 3 above)
Comment (arunthirupathi) [reply presto-orc/src/main/java/com/facebook/presto/orc/OrcWriterOptions.java]: Nice catch, copy/paste mistake :)

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcWriterOptions.java (hunk 4 above)
Comment (arunthirupathi) [reply presto-orc/src/main/java/com/facebook/presto/orc/OrcWriterOptions.java:154]: I will skip it for now.

File: presto-hive/src/main/java/com/facebook/presto/hive/OrcFileWriterConfig.java (hunk 1 above)
Comment (sdruzkin) [reply presto-hive/src/main/java/com/facebook/presto/hive/OrcFileWriterConfig.java]: Isolation and thread safety are side effects of immutability.

File: presto-hive/src/main/java/com/facebook/presto/hive/OrcFileWriterFactory.java (hunk 2 above)
Comment (sdruzkin) [reply presto-hive/src/main/java/com/facebook/presto/hive/OrcFileWriterFactory.java:127]: Are you sure? From my understanding of this change you pass the OrcFileWriterConfig to be able to get the builder later in the `createFileWriter`. If you add a new method `OrcWriterOptions.toBuilder`, that I suggested previously, you will reduce amount of changes in this class and keep exactly the same behavior as it was before this change.

Opinion?

Everything else look good to me.

Comment (sdruzkin) [reply presto-hive/src/main/java/com/facebook/presto/hive/OrcFileWriterFactory.java:127]: I'm sorry, you changed it in the latest commit. LGTM

**Review by sdruzkin — APPROVED** [review]

File: presto-hive/src/main/java/com/facebook/presto/hive/OrcFileWriterFactory.java (hunk 2 above)
Comment (arunthirupathi) [reply presto-hive/src/main/java/com/facebook/presto/hive/OrcFileWriterFactory.java:127]: Discussed offline. The Product code uses Injected constructor which tales in OrcFileWriterConfig. Previously this factory captured the OrcWriterOptions. Now it is capturing the OrcWriterOptions.Builder. 

Previous code, created immutable copies for each with* calls, The new code creates a copy of the builder, modifies all the properties and finally calls build.

**Review by bhhari — APPROVED** [review]

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcWriterOptions.java [hunk 6]
```diff
@@ -316,4 +151,184 @@ public String toString()
...
+                int stripeMaxRowCount,
+                int rowGroupMaxRowCount,
+                DataSize dictionaryMaxMemory,
+                DataSize maxStringStatisticsLimit,
+                DataSize maxCompressionBufferSize,
+                OptionalInt compressionLevel,
+                StreamLayout streamLayout,
+                boolean integerDictionaryEncodingEnabled)
+        {
+            this.stripeMinSize = stripeMinSize;
```
Comment (arhimondr) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcWriterOptions.java:195]: Pleas add the `requireNonNull` checks for everything that cannot be null

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcWriterOptions.java [hunk 7]
```diff
@@ -316,4 +151,184 @@ public String toString()
...
+            return this;
+        }
+
+        public Builder withIntegerDictionaryEncodingEnabled(boolean integerDictionaryEncodingEnabled)
+        {
+            this.integerDictionaryEncodingEnabled = integerDictionaryEncodingEnabled;
+            return this;
+        }
+
+        public DataSize getStripeMinSize()
```
Comment (arhimondr) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcWriterOptions.java:269]: It's a little unusually to have access methods in a builder? Why are they needed?

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcWriterOptions.java [hunk 8]
```diff
@@ -316,4 +151,184 @@ public String toString()
...
+        {
+            return maxStringStatisticsLimit;
+        }
+
+        public DataSize getMaxCompressionBufferSize()
+        {
+            return maxCompressionBufferSize;
+        }
+
+        public OrcWriterOptions.Builder copy()
```
Comment (arhimondr) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcWriterOptions.java:304]: I don't have a strong opinion against it, but it feels like a better interface could be a `OrcWriterOptions#builder(OrcWriterOptions options)` that creates a builder from an existing immutable instance of the `OrcWriterOptions`

File: presto-hive/src/main/java/com/facebook/presto/hive/OrcFileWriterFactory.java [hunk 9]
```diff
@@ -133,7 +133,7 @@ public OrcFileWriterFactory(
         this.nodeVersion = requireNonNull(nodeVersion, "nodeVersion is null");
         this.hiveStorageTimeZone = requireNonNull(hiveStorageTimeZone, "hiveStorageTimeZone is null");
         this.readStats = requireNonNull(readStats, "stats is null");
-        this.orcWriterOptions = requireNonNull(orcWriterOptions, "orcWriterOptions is null");
+        this.orcWriterOptionsBuilder = requireNonNull(orcFileWriterConfig, "orcFileWriterConfig is null").toOrcWriterOptionsBuilder();
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/OrcFileWriterFactory.java:136]: I would recommend avoiding storing a mutable builder as a class field. It could be error prone. How about storing an immutable instance of the `OrcWriterOptions` and create a builder based on that if needed?

File: presto-hive/src/main/java/com/facebook/presto/hive/OrcFileWriterConfig.java [hunk 10]
```diff
@@ -20,94 +20,95 @@
...
 {
-    private OrcWriterOptions options = new OrcWriterOptions();
+    private final OrcWriterOptions.Builder optionsBuilder = OrcWriterOptions.builder();
 
-    public OrcWriterOptions toOrcWriterOptions()
+    public OrcWriterOptions.Builder toOrcWriterOptionsBuilder()
     {
-        return options;
+        // Give separate copy to callers for isolation.
+        return optionsBuilder.copy();
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/OrcFileWriterConfig.java:28]: Oh, now I see why do you need the `copy` and the getters. Basically you are trying to avoid redeclaring all the same fields in the `OrcFileWriterConfig`. But honestly I don't think that's a big deal, and it feels like it could be a reasonable treadoff for a more intuitive builder interface. How strongly do you feel against simply copying the fields into this configuration? (it would also allow you to decouple `OrcWriterOptions` from the Presto specific configs)

File: presto-hive/src/test/java/com/facebook/presto/hive/TestOrcFileWriterConfig.java [hunk 11]
```diff
@@ -65,4 +76,71 @@ public void testExplicitPropertyMappings()
...
+                .setStringStatisticsLimit(new DataSize(17, MEGABYTE))
+                .setMaxCompressionBufferSize(new DataSize(19, MEGABYTE));
+
+        ImmutableList.Builder<SetAndVerifyConfig> builder = ImmutableList.builder();
+        for (int i = 0; i < totalThreads; i++) {
+            builder.add(new SetAndVerifyConfig(config));
+        }
+        ExecutorService executorService = newFixedThreadPool(totalThreads);
+        for (Future<Boolean> result : executorService.invokeAll(builder.build())) {
+            result.get();
```
Comment (arhimondr) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestOrcFileWriterConfig.java:141]: This will cause waiting on the future, so it will be run in a single thread. But honestly it feels like this test is an overkill. Multithreaded testing is hard and often flaky and error prone. If it is possible to avoid it - better to avoid it. But if you feel strongly about this test - feel free to keep it. Just make sure it does what you intended it to do.

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcWriterOptions.java (hunk 7 above)
Comment (arunthirupathi) [reply presto-orc/src/main/java/com/facebook/presto/orc/OrcWriterOptions.java]: getter methods are removed.

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcWriterOptions.java (hunk 8 above)
Comment (arunthirupathi) [reply presto-orc/src/main/java/com/facebook/presto/orc/OrcWriterOptions.java]: this is removed.

File: presto-hive/src/main/java/com/facebook/presto/hive/OrcFileWriterConfig.java (hunk 10 above)
Comment (arunthirupathi) [reply presto-hive/src/main/java/com/facebook/presto/hive/OrcFileWriterConfig.java]: I have copied the fields to the config. This comes with a change in behavior, if OrcFileWriterConfig is mutated, previously OrcFileWriterConfig used to work with the old data before mutation. Now the mutation will be reflected. But looking at the code, it is not mutated once created, so this should be fine.

File: presto-hive/src/main/java/com/facebook/presto/hive/OrcFileWriterFactory.java (hunk 9 above)
Comment (arunthirupathi) [reply presto-hive/src/main/java/com/facebook/presto/hive/OrcFileWriterFactory.java]: Removed the builder.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestOrcFileWriterConfig.java (hunk 11 above)
Comment (arunthirupathi) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestOrcFileWriterConfig.java]: This is my understanding, The threads will run in parallel, but the main thread will wait for them sequentially. So it would produce errors if there are conflicts and I verified before writing the test that it indeed produced conflict. But with the move to the config maintaining its own state, this is not required and I have removed this test.

**Review by arhimondr — APPROVED** [review]
